I've just pushed that on and changed one test:
```diff
diff --git a/sympy/polys/tests/test_rings.py b/sympy/polys/tests/test_rings.py
index 274ac4c702..2d2666c3b1 100644
--- a/sympy/polys/tests/test_rings.py
+++ b/sympy/polys/tests/test_rings.py
@@ -171,7 +171,7 @@ def test_sring():
 
     r = sqrt(2) - sqrt(3)
     R, a = sring(r, extension=True)
-    assert R.domain == QQ.algebraic_field(r)
+    assert R.domain == QQ.algebraic_field(sqrt(2) + sqrt(3))
     assert R.gens == ()
     assert a == R.domain.from_sympy(r)
```
Here the extension element gets normalised by sring.

> the timings almost too good to be true

If you increase `next` the number of extensions then it slows down quite quickly:
```console
$ python time.py    # next=3
n = 1 : 0.080785 secs
n = 2 : 0.0825591 secs
n = 3 : 0.115191 secs
n = 4 : 0.149894 secs
n = 5 : 0.266975 secs
n = 6 : 0.43802 secs
n = 7 : 0.432203 secs
n = 8 : 0.587976 secs
n = 9 : 0.792337 secs

$ python time.py    # next=4
n = 1 : 0.74924 secs
n = 2 : 1.6284 secs
n = 3 : 2.11424 secs
n = 4 : 2.90014 secs
n = 5 : 3.78933 secs
n = 6 : 4.93005 secs
^CTraceback (most recent call last):
  File "time.py", line 16, in <module>
    linsolve(eqs, syms)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solveset.py", line 2645, in linsolve
    sol = solve_lin_sys(eqs, ring, _raw=False)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 238, in solve_lin_sys
    tresult = {to_sympy(sym): to_sympy(val) for sym, val in result.items()}
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 238, in <dictcomp>
    tresult = {to_sympy(sym): to_sympy(val) for sym, val in result.items()}
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 236, in to_sympy
    return ring.domain.to_sympy(x)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domains/algebraicfield.py", line 64, in to_sympy
    return AlgebraicNumber(self.ext, a).as_expr()
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 2532, in as_expr
    return self.as_poly(x or self.root).as_expr().expand()
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3579, in expand
    expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3512, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3512, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3520, in _expand_hint
    newexpr = getattr(expr, hint)(**hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1162, in _eval_expand_multinomial
    return basic_from_dict(expansion_dict, *p)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyutils.py", line 382, in expr_from_dict
    term.append(Pow(g, m))
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 120, in __hash__
    h = self._mhash
KeyboardInterrupt
```
The slow part there seems to converting the solution back to Expr at the end.

--- a/sympy/polys/domainmatrix.py
+++ b/sympy/polys/domainmatrix.py
@@ -253,8 +253,9 @@ def ddm_irref(a):
 
         # normalise row
         ai = a[i]
+        aijinv = aij**-1
         for l in range(j, n):
-            ai[l] /= aij # ai[j] = one
+            ai[l] *= aijinv # ai[j] = one
 
         # eliminate above and below to the right
         for k, ak in enumerate(a):

--- a/sympy/polys/domains/algebraicfield.py
+++ b/sympy/polys/domains/algebraicfield.py
@@ -60,8 +60,11 @@ class AlgebraicField(Field, CharacteristicZero, SimpleDomain):
 
     def to_sympy(self, a):
         """Convert ``a`` to a SymPy object. """
-        from sympy.polys.numberfields import AlgebraicNumber
-        return AlgebraicNumber(self.ext, a).as_expr()
+        # Precompute a converter to be reused:
+        if not hasattr(self, '_converter'):
+            self._converter = _make_converter(self)
+
+        return self._converter(a)
 
     def from_sympy(self, a):
         """Convert SymPy's expression to ``dtype``. """
@@ -137,3 +140,44 @@ class AlgebraicField(Field, CharacteristicZero, SimpleDomain):
     def from_GaussianRationalField(K1, a, K0):
         """Convert a GaussianRational element 'a' to ``dtype``. """
         return K1.from_sympy(K0.to_sympy(a))
+
+
+def _make_converter(K):
+    """Construct the converter to convert back to Expr"""
+    # Precompute the effect of converting to sympy and expanding expressions
+    # like (sqrt(2) + sqrt(3))**2. Asking Expr to do the expansion on every
+    # conversion from K to Expr is slow. Here we compute the expansions for
+    # each power of the generator and collect together the resulting algebraic
+    # terms and the rational coefficients into a matrix.
+    from sympy import Add, S
+
+    gen = K.ext.as_expr()
+    todom = K.dom.from_sympy
+
+    # We'll let Expr compute the expansions. We won't make any presumptions
+    # about what this results in except that it is QQ-linear in some terms
+    # that we will call algebraics. The final result will be expressed in
+    # terms of those.
+    powers = [S.One, gen]
+    for n in range(2, K.mod.degree()):
+        powers.append((gen * powers[-1]).expand())
+
+    # Collect the rational coefficients and algebraic Expr that can
+    # map the ANP coefficients into an expanded sympy expression
+    terms = [dict(t.as_coeff_Mul()[::-1] for t in Add.make_args(p)) for p in powers]
+    algebraics = set().union(*terms)
+    matrix = [[todom(t.get(a, S.Zero)) for t in terms] for a in algebraics]
+
+    # Create a function to do the conversion efficiently:
+
+    def converter(a):
+        """Convert a to Expr using converter"""
+        from sympy import Add, Mul
+        ai = a.rep[::-1]
+        tosympy = K.dom.to_sympy
+        coeffs_dom = [sum(mij*aj for mij, aj in zip(mi, ai)) for mi in matrix]
+        coeffs_sympy = [tosympy(c) for c in coeffs_dom]
+        res = Add(*(Mul(c, a) for c, a in zip(coeffs_sympy, algebraics)))
+        return res
+
+    return converter

--- a/sympy/polys/solvers.py
+++ b/sympy/polys/solvers.py
@@ -367,9 +367,14 @@ def _solve_lin_sys_component(eqs_coeffs, eqs_rhs, ring):
     else:
         sols = {}
         g = ring.gens
-        echelon = echelon.rep
+        # Extract ground domain coefficients and convert to the ring:
+        if hasattr(ring, 'ring'):
+            convert = ring.ring.ground_new
+        else:
+            convert = ring.ground_new
+        echelon = [[convert(eij) for eij in ei] for ei in echelon.rep]
         for i, p in enumerate(pivots):
-            v = echelon[i][-1] - sum(echelon[i][j]*g[j] for j in range(p+1, len(g)))
+            v = echelon[i][-1] - sum(echelon[i][j]*g[j] for j in range(p+1, len(g)) if echelon[i][j])
             sols[keys[p]] = v
 
     return sols

